sdk/source/index.ts

```diff
@@ -60,14 +60,32 @@ export async function getProof(eth_getStorageAt: EthGetStorageAt, eth_getProof:
 	}
 }
 
-export async function getPrice(eth_getStorageAt: EthGetStorageAt, exchangeAddress: bigint, denominationToken: bigint, blockNumber: number): Promise<bigint> {
-	// TODO
-	eth_getStorageAt
-	exchangeAddress
-	denominationToken
-	blockNumber
-
-	return 0n
+export async function getPrice(eth_getStorageAt: EthGetStorageAt, eth_getBlockByNumber: EthGetBlockByNumber, exchangeAddress: bigint, denominationToken: bigint, blockNumber: bigint): Promise<bigint> {
+	async function getAccumulatorValue(blockNumber: bigint | 'latest') {
+		const token0 = uint8ArrayToUnsignedInteger(await eth_getStorageAt(exchangeAddress, 6n, blockNumber))
+		const reservesAndTimestamp = uint8ArrayToUnsignedInteger(await eth_getStorageAt(exchangeAddress, 8n, blockNumber))
+		const accumulator0 = uint8ArrayToUnsignedInteger(await eth_getStorageAt(exchangeAddress, 9n, blockNumber))
+		const accumulator1 = uint8ArrayToUnsignedInteger(await eth_getStorageAt(exchangeAddress, 10n, blockNumber))
+		const blockTimestampLast = reservesAndTimestamp >> (112n + 112n)
+		const reserve1 = (reservesAndTimestamp >> 112n) & (2n**112n - 1n)
+		const reserve0 = reservesAndTimestamp & (2n**112n - 1n)
+		if (reserve0 === 0n) throw new Error(`Exchange ${addressToString(exchangeAddress)} does not have any reserves for token0.`)
+		if (reserve1 === 0n) throw new Error(`Exchange ${addressToString(exchangeAddress)} does not have any reserves for token1.`)
+		if (blockTimestampLast === 0n) throw new Error(`Exchange ${addressToString(exchangeAddress)} has not had its first accumulator update (or it is year 2106).`)
+		if (accumulator0 === 0n) throw new Error(`Exchange ${addressToString(exchangeAddress)} has not had its first accumulator update (or it is 136 years since launch).`)
+		if (accumulator1 === 0n) throw new Error(`Exchange ${addressToString(exchangeAddress)} has not had its first accumulator update (or it is 136 years since launch).`)
+		const numeratorReserve = (token0 === denominationToken) ? reserve0 : reserve1
+		const denominatorReserve = (token0 === denominationToken) ? reserve1 : reserve0
+		const accumulator = (token0 === denominationToken) ? accumulator1 : accumulator0
+		return accumulator + (BigInt(Date.now() / 1000) - blockTimestampLast) * numeratorReserve / denominatorReserve
```

Comment:
Need to scale numeratorReserve  by 2^112

Reply:
Oh and you need to divide num by denominator FIRST before multiplying by seconds. You go out of your way to lose resolution (because you shouldn't have different numbers if called more often)

Reply:
Fixed both.